### PR TITLE
Add BCF reader/writer and utility functions for VCF.

### PR DESCRIPTION
**Pull request by alumi**
This PR adds reader and writer for [BCF v2.2 file format](https://samtools.github.io/hts-specs/VCFv4.3.pdf) which is a binary version of VCF.

We can read/write a BCF file using `cljam.bcf`.

```clojure
(with-open [r (cljam.bcf/reader "/PATH/TO/BCF.bcf")]
  (doall (cljam.bcf/read-variants r)))
=> ({:chrom "20", :pos 1234567, :id "microsat1", ...} ...)

(with-open [w (cljam.bcf/writer "/PATH/TO/BCF.bcf")]
  (cljam.bcf/write-variants w [{:chrom "20", :pos 1234567, :id "microsat1", ...} ...]))
```

Included test file `test-v4_3.bcf` is converted with `bcftools view -O b test-v4_3.vcf`.

### about variant representation
Currently, we use following map to represent a variant of VCF.

##### VCF
```
20      1234567 microsat1       GTC     G,GTCT  50      PASS    NS=3;DP=9;AA=G  GT:GQ:DP        0/1:35:4        0/2:17:2        1/1:40:3
```

##### cljam.vcf-style variant map
```clojure
{:chrom "20", :pos 1234567, :id "microsat1", :ref "GTC", :alt ["G" "GTCT"], :qual 50.0,
 :filter "PASS", :info "NS=3;DP=9;AA=G", :FORMAT "GT:GQ:DP",
 :NA00001 "0/1:35:4", :NA00002 "0/2:17:2", :NA00003 "1/1:40:3"}
```

Since BCF file format requires type information, we need to parse FILTER, INFO, FORMAT and samples columns.
So I introduced parsed version of variant representation.

##### parsed variant map
```clojure
{:chr "20", :pos 1234567, :id "microsat1", :ref "GTC", :alt ["G" "GTCT"], :qual 50.0,
 :filter [:PASS], :info {:NS 3, :DP 9, :AA "G"}, :FORMAT [:GT :GQ :DP],
 :NA00001 {:GT "0/1", :GQ 35, :DP 4}, :NA00002 {:GT "0/2", :GQ 17, :DP 2}, :NA00003 {:GT "1/1", :GQ 40, :DP 3}}
```

Functions for conversion of VCF style map <=> parsed style map are in `cljam.util.vcf-util`.

Then I introduced another version of variant representation for intermediate format of BCF.
It contains indices to meta-info of the file.

##### cljam.bcf-style variant map
```clojure
{:chr 0, :pos 1234567, :id "microsat1", :ref-length 3, :ref "GTC", :alt ["G" "GTCT"], :qual 50.0,
 :filter [0],  :info [[1 [3]] [2 [9]] [4 "G"]], :n-sample 3,
 :genotype [[9 [[2 4] [2 6] [4 4]]] [10 [[35] [17] [40]]] [2 [[4] [2] [3]]]]}
```

Finally, BCF-style map can be converted to/from BCF.

##### BCF
```
3b 00 00 00 15 00 00 00  00 00 00 00 86 d6 12 00  |;...............|
03 00 00 00 00 00 48 42  03 00 03 00 03 00 00 03  |......HB........|
97 6d 69 63 72 6f 73 61  74 31 37 47 54 43 17 47  |.microsat17GTC.G|
47 47 54 43 54 11 00 11  01 11 03 11 02 11 09 11  |GGTCT...........|
04 17 47 11 09 21 02 04  02 06 04 04 11 0a 11 23  |..G..!.........#|
11 28 11 02 11 04 02 03                           |.(......|
```

Thus, their convertibilities can be illustrated as follows.

```
            VCF-style map --- VCF (tsv)
          /
parsed map
          \
            BCF-style map --- BCF (binary)
```

In this PR, 
- ~~~`cljam.vcf` uses VCF-style maps~~~
- ~~~`cljam.bcf` converts BCF <-> BCF-style maps <-> parsed maps <-> VCF-style maps~~~

~~~I'm wondering if we should use parsed maps as default for both of `cljam.vcf` and `cljam.bcf`.~~~

- `cljam.vcf` and `cljam.bcf` returns fully parsed maps.
- `read-variants` accepts option `:depth` to specify format of returned values.
- The key `:chrom` is renamed to `:chr`

### CODE REVIEW

**Comment by totakke** [comment]
Is this ready to be reviewed? If so, please remove `[WIP]` from PR title.

**Comment by alumi** [comment]
@totakke Thank you for discussion about the spec!
I've added some more commits.

File: test/cljam/util/t_vcf_util.clj [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+      "HOMLEN=1" {:HOMLEN [1]}
+      "HOMLEN=1,2,3,4,5" {:HOMLEN [1 2 3 4 5]}
+      "CC=A" {:CC \A}
+      "DB"   {:DB :exists}
+      "DB;H2" {:DB :exists :H2 :exists}
+      "NS=3;DP=13;AA=T" {:NS 3, :DP 13, :AA "T"}
+      "NS=3;DP=9;AA=G" {:NS 3, :DP 9, :AA "G"}
+      "NS=3;DP=14;AF=0.5;DB;H2" {:NS 3, :DP 14, :AF [0.5], :DB :exists :H2 :exists}
+      "NS=3;DP=11;AF=0.017" {:NS 3, :DP 11, :AF [(Float/parseFloat "0.017")]}
+      "NS=2;DP=10;AF=0.333,0.667;AA=T;DB" {:NS 2, :DP 10, :AF [(Float/parseFloat "0.333") (Float/parseFloat "0.667")] :AA "T", :DB :exists})))
```
Comment (totakke) [inline test/cljam/util/t_vcf_util.clj]: `(float 0.333)` is more simple.

File: src/cljam/bcf.clj [hunk 2]
```diff
@@ -0,0 +1,32 @@
+(ns cljam.bcf
```
Comment (totakke) [inline src/cljam/bcf.clj]: Please add docstring of the namespace.

**Comment by alumi** [comment]
@totakke Thank you for your advice. I fixed the wrapping, docstring, typehints and float constants.

**Comment by totakke** [comment]
Thank you. Good job.

**Pull request merged**